powershell-adapter/powershell.resource.ps1

```diff
@@ -1,427 +1,349 @@
-                # so we just need to signal error to the caller by non-zero exit code.
+            Default {
+                $errmsg = 'Can not find implementation of type: "' + $cachedResourceInfo.ImplementationDetail + '". If this is a binary resource such as File, use the Microsoft.Dsc/WindowsPowerShell adapter.'
+                Write-Error $errmsg
                 exit 1
             }
-            $result = $op_result
         }
-        else
-        {
-            $errmsg = "Can not find type " + $inputobj_pscustomobj.type + "; please ensure that Get-DscResource returns this resource type"
-            Write-Error $errmsg
-            exit 1
-        }
-    }
 
-    $result | ConvertTo-Json
+        return $addToActualState
+    }
+    else {
+        $errmsg = 'Can not find type "' + $ds.type + '". Please ensure that Get-DscResource returns this resource type.'
+        Write-Error $errmsg
+        exit 1
+    }
 }
-elseif ($Operation -eq 'Export')
-{
-    $result = @()
-
-    RefreshCache
-
-    if (IsConfiguration $inputobj_pscustomobj) # we are processing a config batch
-    {
-        foreach($r in $inputobj_pscustomobj.resources)
-        {
-            $cachedResourceInfo = $script:ResourceCache[$r.type]
-            if ($cachedResourceInfo)
-            {
-                $path = $cachedResourceInfo.Path # for class-based resources - this is path to psd1 of their defining module
 
-                $typeparts = $r.type -split "/"
-                $ResourceTypeName = $typeparts[1]
+# Get-TypeInstanceFromModule function to get the type instance from the module
+function Get-TypeInstanceFromModule {
+    param(
+        [Parameter(Mandatory = $true)]
+        [string] $modulename,
+        [Parameter(Mandatory = $true)]
+        [string] $classname
+    )
+    $instance = & (Import-Module $modulename -PassThru) ([scriptblock]::Create("'$classname' -as 'type'"))
+    return $instance
+}
 
-                $scriptBody = "using module '$path'"
-                $script = [ScriptBlock]::Create($scriptBody)
-                . $script
+# initialize OUTPUT as array
+$result = [System.Collections.Generic.List[Object]]::new()
+
+# process the operation requested to the script
+switch ($Operation) {
+    'List' {
+        # cache was refreshed on script load
+        foreach ($Type in $resourceCache.Type) {
+        
+            # https://learn.microsoft.com/dotnet/api/system.management.automation.dscresourceinfo
+            $r = $resourceCache | Where-Object Type -EQ $Type | ForEach-Object DscResourceInfo
+
+            # Provide a way for existing resources to specify their capabilities, or default to Get, Set, Test
+            $module = Get-Module -Name $r.ModuleName -ListAvailable | Sort-Object -Property Version -Descending | Select-Object -First 1
+            if ($module.PrivateData.PSData.Capabilities) {
+                $capabilities = $module.PrivateData.PSData.Capabilities
```

Comment:
Perhaps we should call it `DscCapabilities` to make it clear it's only for DSC?

Reply:
fixed